Implement and with it, improve logging

MKodde opened the pull request:
The PSR-3 logger interface is injected in the most important services of the library. An effort was put into implementing actual log messages in those services. A good review  is in order here. As I analyzed the code and added log messages in places I found usefull. But feel free to suggest other situations where logging is important. Also please review the relevance and correctness of the log messages themselves.

https://www.pivotaltracker.com/story/show/181558157

MKodde committed:
Add PSR-3 support, allowing LoggerInterface usage

MKodde committed:
Implement logging in OcraService

The factory now injects the logger interface on the implemtations it is
able to create. Logging was applied on a quick win basis in the
implmentations.

MKodde committed:
Add logging to DeviceStorage

MKodde committed:
Add logging to StateStorage implementations

MKodde committed:
Add logging to User(Secret)Storage implementations

MKodde committed:
Add logging to the Tiqr_Service

MKodde committed:
Update tests

MKodde committed:
Improve README.md adding logging instructions

MKodde committed:
Remove LDAP support from User(Secret)Storage

Removed support for creating an instance of a custom storage solution
by providing the FQN of the implementation as $type. This construction
makes the library more open for implementation. If ever somebody wants
to support a different storage solution, they can add it to the library,
by creating a concrete solution to the library + updating the factory.

MKodde committed:
Remove support for 'magic' factory feature

Creating instances of types provided by giving a FQN has been removed
from the other factory methods, as was already applied for the user
storage solution.

MKodde committed:
Increase test coverage for factories

Testing correct creation, and also testing that creating a non existing
service results in a runtime exception.

Also removed a stray break operator in this commit. Sorry for breaking
the atomic commit rule on this one.

MKodde committed:
Add .phpunit.result.cache to .gitignore

MKodde committed:
Enable PHPUnit code coverage on Travis CI

MKodde committed:
Test UserStorage_File for refactoring

pmeulen commented on library/tiqr/Tiqr/DeviceStorage/TokenExchange.php:
```diff
@@ -44,8 +44,16 @@ public function getDeviceToken($notificationToken)
         $output = file_get_contents($url);
-        if (stripos($output, "not found")!==false) return false;
-        if (stripos($output, "error")!==false) return false;
+        if (stripos($output, "not found")!==false) {
```
Is this way to check for error correct? I understand file_get_contents() returns false on error (https://www.php.net/manual/en/function.file-get-contents.php).

Edit: OK I see it now :) The is the actual response from the token exchange. Asserting that $output !== false is a good check to add IMO.

Not your code, but finding a way to handle any "not OK" response as an error would make more sense instead of checking for possible errors here. That might be hard to do. What we can do is add tracing with loglevel e.g. INFO, so whatever the response of the token exchange sent will be available.

MKodde replied:
Yes! Added a notice log with the output when no error occurred.

MKodde committed:
Move Encryption implementation > UserSecretStorage

They are only used by the UserSecretStorage solution, so moving them to
that 'namespace' makes perfect sense

MKodde committed:
Create the UserSecretStorageTrait

This trait will be used by the User Secret Storage implementations in
favour of using an abstract implementation. As we do not particularly
like to commit to inheritance. See the previous mess this gotten us in
to.

pmeulen commented on library/tiqr/Tiqr/OcraService/OathServiceClient.php:
```diff
@@ -30,29 +32,36 @@ class Tiqr_OcraService_OathServiceClient extends Tiqr_OcraService_Abstract
         if ($result->code == '200') {
             return $result->body;
         }
+        $this->logger->error('The call to /oath/challenge/ocra did not yield a challenge.');
```
Add a NOTICE with the result code and body?

MKodde replied:
Surething! Added this for `generateChallenge` & `verifyResponseWithUserId`

pmeulen commented on library/tiqr/Tiqr/UserSecretStorage/OathServiceClient.php:
```diff
@@ -31,8 +33,9 @@ class Tiqr_UserSecretStorage_OathServiceClient implements Tiqr_UserSecretStorage
-    public function __construct($config)
+    public function __construct($config, LoggerInterface $logger, $secretconfig = array())
     {
+        $this->logger = $logger;
```
Srutenizer says: "The property logger does not exist. Although not strictly required by PHP, it is generally a best practice to declare properties explicitly."

I agree.

MKodde replied:
I already tackled that on one of the other PR's. Consider this fixed!

pmeulen commented on library/tiqr/Tiqr/UserSecretStorage/Pdo.php:
```diff
@@ -36,13 +38,16 @@ class Tiqr_UserSecretStorage_Pdo extends Tiqr_UserStorage_Pdo implements Tiqr_Us
         try {
             $this->handle = new PDO($config['dsn'],$config['username'],$config['password']);
         } catch (PDOException $e) {
-            return false;
```
Do we want to update the phpdoc of the functions that now throw instead of returning false?

MKodde replied:
My personal preference is to remove all PHPDoc that contributes no valuable explanation.

Some methods now throw an exception. And that might be a good thing to advertise on the PHPDoc!

pmeulen commented on library/tiqr/Tiqr/UserStorage/Pdo.php:
```diff
@@ -57,61 +61,87 @@ public function createUser($userId, $displayName)
+     * Be carefull! This method does not return an expected boolean value. Instead it returns:
+     * - void (if user was not found)
+     * - userid (if user was found)
+     *
```
Yes, and this is inconsistent with its parent's and and siblings interfaces. We should fix this and define the same interface for all.

Adding a TODO: here and not resolving it in the PR is fine here. There is a limit in what you can/should do in one PR and I like you are adding notes where you find code that needs further attention.

MKodde committed:
Remove UserStorage 'secret' logic

MKodde committed:
Integrage 'secret' logic in the UserSecretStorage

This was previously intermingled with the regular user storage

VadimSchmitz commented on library/tiqr/Tiqr/OcraService/OathServiceClient.php:
```diff
@@ -30,29 +32,36 @@ class Tiqr_OcraService_OathServiceClient extends Tiqr_OcraService_Abstract
      * Get the ocra challenge
      *
-     * @return String The challenge
+     * @return string|null The challenge or null when no challende was returned form the Oath service
```
challange typo

VadimSchmitz commented on library/tiqr/Tiqr/UserSecretStorage.php:
```diff
@@ -36,32 +38,32 @@ class Tiqr_UserSecretStorage
                 break;
             case "ldap":
                 require_once("Tiqr/UserSecretStorage/Ldap.php");
                 $instance = new Tiqr_UserSecretStorage_Ldap($options);
```
Why does Tiqr_UserSecretStorage_Ldap not have a logger?

MKodde replied:
You will see that in one of the next PR's I'm going to remove the LDAP implementations. No use adding logging IMHO.

Well spotted by the way :+1:

VadimSchmitz commented on library/tiqr/Tiqr/UserStorage/Pdo.php:
```diff
@@ -57,61 +61,87 @@ public function createUser($userId, $displayName)
+        if ($sth->execute(array($userId))) {
+            return $sth->fetchColumn();
+        }
+        $this->logger->error('Unable fot find user in user storage (PDO)');
```
typo?

pmeulen commented on library/tiqr/Tiqr/UserStorage/File.php:
```diff
@@ -34,9 +36,9 @@ class Tiqr_UserStorage_File extends Tiqr_UserStorage_GenericStore
      */
-    public function __construct($config, $secretconfig = array())
+    public function __construct($config, LoggerInterface $logger, $secretconfig = array())
     {
```
PHPDOC needs updating

pmeulen commented on library/tiqr/Tiqr/DeviceStorage.php:
```diff
@@ -42,16 +44,16 @@ class Tiqr_DeviceStorage
      * @throws Exception An exception if an unknown storage is requested.
      */
-    public static function getStorage($type="dummy", $options=array())
+    public static function getStorage($type="dummy", $options=array(), LoggerInterface $logger)
```
PHPDOC needs updating.

pmeulen reviewed (changes requested):
Nice work!

The error logging and tracing you added should be a big help in diagnosing problems. We'll see how this logging affects the log volume when we use the library and adjust when needed. Stepup-tiqr uses fingers-crossed logging and normally only logs NOTICE and above, that should help a lot.

When working on code like this it is expected that you come across some broken windows. I think it is useful to add TODO: for these.

Lot's of interfaces changed and not all PHPDOC sections were updated accordingly. We should strive to update at least the PHPDOC of public API op the library IMO.

MKodde committed:
Update PHPDoc

Adding missing @throws and @param

MKodde committed:
Log OcraServiceClient api response

MKodde committed:
Add logging to happy flow of TokenExchange

MKodde committed:
Addressed some typos in log message and PHPDoc

MKodde commented:
Thanks for the reviews @VadimSchmitz and @pmeulen Commits: 7c94eba..296a6c5 addressed the changes mentioned in the feedback.

The large amount of additional commits are from the other PR's that landed in this branch.

MKodde committed:
Add the logger field on user secret storage Pdo
